pyqtorch/modules/abstract.py

```diff
@@ -15,21 +16,26 @@ def __init__(self, qubits: ArrayLike, n_qubits: int):
         self.qubits = qubits
         self.n_qubits = n_qubits
 
-    def __mul__(
-        self, other: AbstractGate | pyqtorch.modules.QuantumCircuit
-    ) -> pyqtorch.modules.QuantumCircuit:
+    def __mul__(self, other: AbstractGate | pyq.QuantumCircuit) -> pyq.QuantumCircuit:
         if isinstance(other, AbstractGate):
-            return pyqtorch.modules.QuantumCircuit(
-                max(self.n_qubits, other.n_qubits), [self, other]
-            )
+            ml = torch.nn.ModuleList([self, other])
+            return pyq.QuantumCircuit(max(self.n_qubits, other.n_qubits), ml)
+        elif isinstance(other, pyq.QuantumCircuit):
+            ml = torch.nn.ModuleList([self]) + other.operations
+            return pyq.QuantumCircuit(max(self.n_qubits, other.n_qubits), ml)
+        else:
+            return ValueError(f"Cannot compose {type(self)} with {type(other)}")
 
-        if isinstance(other, pyqtorch.modules.QuantumCircuit):
-            return pyqtorch.modules.QuantumCircuit(
-                max(self.n_qubits, other.n_qubits), [self, *other.operations]
-            )
+    def __key(self) -> tuple:
+        return (self.n_qubits, *self.qubits)
 
-        else:
-            return NotImplemented
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, type(self)):
```

Comment:
`AbstractGate` ? Unless this equality checking works for subtypes ?

Reply:
This would result in `X==Y` returning true